### ISSUE TITLE
SignalR authorization not working in asp.net core angular SPA with Identity

### ISSUE THREAD

**Issue by maxima120** [dotnet/aspnetcore#34948]
### Describe the bug
When using Identity Server authorization of a SignalR hub on server side fails. Auth for controllers works fine.

**Github complete code**: https://github.com/maxima120/signalr-test
*the problem occured in a prod app which is quite mature, I just need to add signalr auth. the code above is a quick repro out of the box Microsoft template without any particular logic.*

### To Reproduce
Created brand new asp.net core 5 angular spa app form template:
```
dotnet new angular --auth Individual
npm i @microsoft/signalr
```

Modified Startup.cs
```
            services.AddCors(options =>
            {
                options.AddPolicy("CorsPolicy", builder => builder
                .WithOrigins("http://localhost:4200")
                .AllowAnyMethod()
                .AllowAnyHeader()
                .AllowCredentials());
            });
            services.AddSignalR();

    . . .

            app.UseCors("CorsPolicy");

            app.UseAuthentication();
            app.UseIdentityServer();
            app.UseAuthorization();
            app.UseEndpoints(endpoints =>
            {
                . . .
                endpoints.MapHub<NewsHub>("/newshub");
            });
```

Added Hub class
```
    [Authorize]
    public class NewsHub : Hub
    {
    }
```

Modified WeatherForecastController:
```
        private IHubContext<NewsHub> _hub;

        public WeatherForecastController(ILogger<WeatherForecastController> logger, IHubContext<NewsHub> hub)
        {
            _hub = hub;
            _logger = logger;
        }

        [HttpGet]
        public IEnumerable<WeatherForecast> Get()
        {
            var timerManager = new TimerManager(() => 
                _hub.Clients.All.SendAsync("servermessage", DateTime.Now.Ticks.ToString()));
```

Modify fetch-data.component.ts
```
    constructor(http: HttpClient, @Inject('BASE_URL') baseUrl: string) {
      http.get<WeatherForecast[]>(baseUrl + 'weatherforecast').subscribe(result => {

      this.forecasts = result;

      this.hub = new HubConnectionBuilder()
        .withUrl("/newshub")
        .build();

      this.hub.on("servermessage", (m: string) => { console.log(m); });

      this.hub.start()
        .then(() => console.log('MessageHub Connected'))
        .catch(err => console.log('MessageHub Connection Error: ' + err.toString()));

    }, error => console.error(error));
  }
```

Authorization of SignalR hub fails. Output window:
```
IdentityServer4.Hosting.IdentityServerMiddleware: Information: Invoking IdentityServer endpoint: IdentityServer4.Endpoints.DiscoveryEndpoint for /.well-known/openid-configuration
IdentityServer4.Hosting.IdentityServerMiddleware: Information: Invoking IdentityServer endpoint: IdentityServer4.Endpoints.DiscoveryEndpoint for /.well-known/openid-configuration
IdentityServer4.Hosting.IdentityServerMiddleware: Information: Invoking IdentityServer endpoint: IdentityServer4.Endpoints.UserInfoEndpoint for /connect/userinfo
IdentityServer4.ResponseHandling.UserInfoResponseGenerator: Information: Profile service returned the following claim types: sub preferred_username name
IdentityServer4.Hosting.IdentityServerMiddleware: Information: Invoking IdentityServer endpoint: IdentityServer4.Endpoints.CheckSessionEndpoint for /connect/checksession
[2021-08-01T15:43:11.337Z] Information: Normalizing '/newshub' to 'https://localhost:44306/newshub'.
Failed to load resource: the server responded with a status of 401 () [https://localhost:44306/newshub/negotiate?negotiateVersion=1]
[2021-08-01T15:43:11.347Z] Error: Failed to complete negotiation with the server: Error
[2021-08-01T15:43:11.347Z] Error: Failed to start the connection: Error
MessageHub Connection Error: Error
```

If I remove [Authorize] attribute - it works fine

I did read the article https://docs.microsoft.com/en-us/aspnet/core/signalr/authn-and-authz?view=aspnetcore-5.0
and I tried to implement suggestion in: "Identity Server JWT authentication" paragraph (PostConfigure) but I understand this requires Identity Server token passed from client to server using `accessTokenFactory`. I spent 2 whole working days trying to figure out how do I extract server token in angular and I give up.. 

I believe this should work out of the box (the signalr auth), same way as the controllers auth (which IS working out of the box). So I call it a bug. Please help.

### Further technical details
Windows 10 Pro, Visual Studio 2019 Pro, Edge Browser 92 x64

.NET SDK (reflecting any global.json):                                                                                                                                               Version:   5.0.302
 Commit:    c005824e35

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.302\

Host (useful for support):
  Version: 5.0.8
  Commit:  35964c9215

.NET SDKs installed:
  5.0.102 [C:\Program Files\dotnet\sdk]
  5.0.302 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

**Comment by maxima120** [comment]
Thank you Brennan. Of course I am not expecting you to give app support. So Ill just ask these to make sure I understand.
You are confident that 
- The code I have is fair implementation and all is left - to extract server token and pass it back;
- SignalR authorization is not supposed to work with Identity Server out of the box and in order to make it work I need a manual implementation of authorization, bypassing existing Identity Server solution

if the latter is correct - can I pass a suggestion that the SignalR team do something about it - because its not an exotic requirement. I believe that Identity Server (of any flavour) is preferred solution for authorization. So why not make SignalR friendly to that (ie configuring SignalR auth should be as simple as all the rest of asp.net core app auth (eg controllers), with few lines in startup and an [Authorize] attribute).

**Comment by BrennanConroy** [comment]
>     * all is left - to extract server token and pass it back

Probably, I can't be 100% confident though, every app is different.

>     * not supposed to work with Identity Server out of the box

I wouldn't say that. It doesn't work "out of the box" with angular because you need to pass the token to SignalR.

> why not make SignalR friendly to that (ie configuring SignalR auth should be as simple as all the rest of asp.net core app auth (eg controllers)

Unfortunately it isn't trivial to make SignalR as simple as an HTTP request. It looks like Angular goes out of its way to make making an authorized HTTP request super simple, but SignalR is more than just an HTTP request, and Angular doesn't have any knowledge about SignalR, so there isn't any nice built in "authorized SignalR" feature that Angular provides. If you have any suggestions to make the docs easier to consume, we'd love to hear it on the [docs repo](https://github.com/dotnet/AspNetCore.Docs)

**Comment by maxima120** [comment]
I got it. It seems a lot of work though. So I came up with a workaround (see the SO link and/or Girhub code) - I send `hub.connectionId` through authorized api call and store userId against it. Then I can cherry pick sending messages from server using `hub.Cleints.Client(MycconnectionsCache[0]).SendAsync`.

**Comment by BrennanConroy** [comment]
> It seems a lot of work though

I don't see how it's that difficult, you have a function for getting the access token, you just need to call it and pass the token to SignalR.
There are tutorials online about how to get the access token https://code-maze.com/secure-angular-calls-to-webapi-using-access-token/

**Comment by maxima120** [comment]
I got through bunch of them in past 3 days. including this one. they all more or less require to write authorization service on the client (Angular). I was hoping to find more "natural" solution but after you offered me essentially the same approach yesterday I realised that there is nothing natural about this issue so I thought up a workaround (which works and afford me decent level of protection).

Now why I dont want to copy paste some lengthy code from blogs on the internet? Thats different question. I dont want to be responsible for security flaws and it is still a lot of work because again - everything works fine now. No code, just few lines in Startup.cs.. The "new" approach will have me writing different solution for an app which has been in prod for a bit. And for what? To make signalr authorize attribute work subject to multiple bugs and holes in the whole use of Identity Server for the app.. I decided not to.
